docs/devel/api_changes.md

```diff
@@ -0,0 +1,231 @@
+   * does not change existing semantics
+   * does not change existing defaults
+
+Put another way:
+
+1. An API call (e.g. a structure POSTed to a REST endpoint) that worked before
+   your change must work the same after your change.
+2. An API call that uses your change must work correctly against servers that
+   do not include your change.
+3. It must be possible to round-trip you change between versioned APIs with no
+   loss of information.
+
+If your change does not meet these criteria, it is not considered strictly
+compatible.  There are times when this might be OK, but mostly we want changes
+that meet this definition.
+
+Let's consider some examples.  In a hypothetical API, the `Frobber` struct
+looks something like this:
+
+```go
+type Frobber struct {
+	Height int    `json:"height"`
+	Param  string `json:"param"`
+}
+```
+
+You want to add a new `Width` field.  It is generally safe to add new fields,
+so you can simply change it to:
+
+```go
+type Frobber struct {
+	Height int    `json:"height"`
+	Width  int    `json:"width"`
+	Param  string `json:"param"`
+}
+```
+
+The onus is on you to define a sane default value for `Width` such that rule #1
+above is true - API calls that used to work must continue to work.
+
+For your next change you want to allow multiple `Param` values.  You can not
+simply change `Param string` to `Params []string` - that fails rules #1 and #2.
+You can instead do something like:
+
+```go
+type Frobber struct {
+	Height int           `json:"height"`
+	Width  int           `json:"width"`
+	Param  string        `json:"param"`  // the first param
+	ExtraParams []string `json:"params"` // additional params
+}
+```
+
+Now you can satisfy the rules: API calls that provide the old style `Param`
+will still work, while servers that don't understand `ExtraParams` can ignore
+it.  This is somewhat unsatisfying as an API, but it is strictly compatible.
+
+Part of the reason for versioning APIs and for using an explicit internal API
+is to handle growth like this.  The internal API can be implemented as:
+
+```go
+type Frobber struct {
+	Height int           `json:"height"`
+	Width  int           `json:"width"`
+	Params []string      `json:"params"`
+}
+```
+
+The code that converts to/from versioned APIs can decode this into the somewhat
+uglier (but compatible!) structures.  Eventually, a new API version will be
+forked and it can use the internal structure.
+
+We've seen how to satisfy rules #1 and #2.  Rule #3 means that you can not
+extend one versioned API without also extending the others.  For example, an
+API call might POST an object in API v7beta1 format, which uses the cleaner
+`Params` field, but the API server might store that object in v6 form (since
+v7beta1 is "beta").  When the user reads the object back in the v7beta1 API it
+would be unacceptable to have lost all but `Params[0]`.  This means that, even
+though it is ugly, a compatible change must be made to the v6 API.
```

Comment:
I think I understand what you're trying to say here, but it's not at all clear.  I'd suggest making clear which example above is v6 and which is v7beta1 (the last two, I assume), and what change you would expect to be made to v6.